proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,374 @@
+
+When compiling with binary stability mode on, a struct can be marked `@frozen` as long as it meets all of the following conditions:
+
+- The struct is *ABI-public* (see [SE-0193][]), i.e. `public` or marked `@abiPublic`.
+- Every class, enum, struct, protocol, or typealias mentioned in the types of the struct's fields is ABI-public.
+- No fields have observing accessors (`willSet` or `didSet`).
+- If a field has an initial value, the expression computing the initial value does not reference any types or functions that are not ABI-public.
+
+```swift
+@frozen
+public struct Point {
+  public var x: Double
+  public var y: Double
+
+  public init(_ x: Double, _ y: Double) {
+    self.x = x
+    self.y = y
+  }
+}
+```
+
+This affects what changes to the struct's fields affect the ABI of the containing library:
+
+| Change | Normal struct | `@frozen` struct
+|---|:---:|:---:
+| Adding fields | Allowed | **Affects ABI**
+| Reordering fields | Allowed | **Affects ABI**
+| Removing ABI-public fields | Affects ABI | Affects ABI
+| Removing non-ABI-public fields | Allowed | **Affects ABI**
+| Changing the type of an ABI-public field | Affects ABI | Affects ABI
+| Changing the type of a non-ABI-public field | Allowed | **Affects ABI**
+| Changing a stored instance property to computed | Allowed | **Affects ABI**
+| Changing a computed instance property to stored | Allowed | **Affects ABI**
+| Changing the access of a non-ABI-public field | Allowed | Allowed
+| Marking an `internal` field as `@abiPublic` | Allowed | Allowed
+| Changing an `internal` ABI-public field to be `public` | Allowed | Allowed
+
+
+> Note: This proposal is implemented already and in use by the standard library, albeit under different names. The command-line flag is `-enable-resilience`; the attribute is `@_fixed_layout` for structs, and `@frozen` for enums.
+
+  [SE-0193]: https://github.com/apple/swift-evolution/blob/master/proposals/0193-cross-module-inlining-and-specialization.md
+
+#### Guarantees
+
+Marking a struct `@frozen` only guarantees that its stored instance properties won't change. This allows the compiler to perform certain optimizations, like ignoring properties that are never accessed, or eliminating redundant loads from the same instance property. However, it does not provide a handful of other guarantees that a C struct might:
+
+- **It is not guaranteed to be "trivial"** ([in the C++ sense][trivial]). A frozen struct containing a class reference or closure still requires reference-counting when copied and when it goes out of scope.
+
+- **It is not guaranteed to be stored in registers.** Promoting structs to be stored in registers is considered an optimization in Swift, and as such the compiler is never *required* to do so. In particular, a struct with a `weak` reference in it must always be stored in memory if the compiler cannot prove that the weak reference is unused.
```

Comment:
This isn't an ABI guarantee, so in hindsight I'm not sure it's worth including. (I'm pretty sure I was the one who added it originally.)